Website/Providers/DataProviders/SqlDataProvider/09.04.00.01.SqlDataProvider

```diff
@@ -0,0 +1,168 @@
+        U.LastIPAddress,
+        U.UpdatePassword,
+        U.PasswordResetToken,
+        U.PasswordResetExpiration,
+        UP.Authorised,
+        UP.HasAgreedToTerms,
+        UP.HasAgreedToTermsOn,
+        U.CreatedByUserId,
+        U.CreatedOnDate,
+        U.LastModifiedByUserId,
+        U.LastModifiedOnDate
+	FROM       {databaseOwner}[{objectQualifier}Users]       AS U
+		LEFT JOIN {databaseOwner}[{objectQualifier}UserPortals] AS UP 
+			ON CASE WHEN U.IsSuperuser = 1 THEN 0 ELSE U.UserId END = UP.UserId
+GO
+
+IF  EXISTS (SELECT * FROM sys.objects WHERE object_id = OBJECT_ID(N'{databaseOwner}[{objectQualifier}GetUsersByRolename]') AND type in (N'P', N'PC'))
+	DROP PROCEDURE {databaseOwner}[{objectQualifier}GetUsersByRolename]
+GO
+
+CREATE PROCEDURE {databaseOwner}[{objectQualifier}GetUsersByRolename]
+	@PortalID	INT,
+	@Rolename	NVARCHAR(50)
+AS
+	DECLARE @UserPortalId INT
+	DECLARE @PortalGroupId INT
+	SELECT @PortalGroupId = PortalGroupId FROM {databaseOwner}[{objectQualifier}Portals] WHERE PortalID = @PortalID
+	IF EXISTS(SELECT PortalGroupID FROM {databaseOwner}[{objectQualifier}PortalGroups] WHERE PortalGroupID = @PortalGroupId)
+	BEGIN
+		SELECT @UserPortalId = MasterPortalID FROM {databaseOwner}[{objectQualifier}PortalGroups] WHERE PortalGroupID = @PortalGroupId
+	END
+	ELSE
+	BEGIN
+		SELECT @UserPortalId = @PortalID
+	END
+	SELECT     
+		U.*, 
+		UP.PortalId, 
+		UP.Authorised, 
+    UP.HasAgreedToTerms,
+    UP.HasAgreedToTermsOn,
+		UP.IsDeleted,
+		UP.RefreshRoles,
+		UP.VanityUrl
+	FROM {databaseOwner}{objectQualifier}UserPortals AS UP 
+			RIGHT OUTER JOIN {databaseOwner}{objectQualifier}UserRoles  UR 
+			INNER JOIN {databaseOwner}{objectQualifier}Roles R ON UR.RoleID = R.RoleID 
+			RIGHT OUTER JOIN {databaseOwner}{objectQualifier}Users AS U ON UR.UserID = U.UserID 
+		ON UP.UserId = U.UserID	
+	WHERE ( UP.PortalId = @UserPortalId OR @UserPortalId IS Null )
+		AND (UP.IsDeleted = 0 OR UP.IsDeleted Is NULL)
+		AND (R.RoleName = @Rolename)
+		AND (R.PortalId = @PortalID OR @PortalID IS Null )
+	ORDER BY U.FirstName + ' ' + U.LastName
+
+GO
+
+IF  EXISTS (SELECT * FROM sys.objects WHERE object_id = OBJECT_ID(N'{databaseOwner}[{objectQualifier}UserAgreedToTerms]') AND type in (N'P', N'PC'))
+	DROP PROCEDURE {databaseOwner}[{objectQualifier}UserAgreedToTerms]
+GO
+
+CREATE PROCEDURE {databaseOwner}[{objectQualifier}UserAgreedToTerms]
+	@PortalId	INT,
+	@UserId		INT
+AS
+	UPDATE {databaseOwner}{objectQualifier}UserPortals
+	SET HasAgreedToTerms = 1, HasAgreedToTermsOn = GETDATE()
+	WHERE PortalId = @PortalId AND UserId = @UserId
+GO
+
+IF  EXISTS (SELECT * FROM sys.objects WHERE object_id = OBJECT_ID(N'{databaseOwner}[{objectQualifier}ResetTermsAgreement]') AND type in (N'P', N'PC'))
+	DROP PROCEDURE {databaseOwner}[{objectQualifier}ResetTermsAgreement]
+GO
+
+CREATE PROCEDURE {databaseOwner}[{objectQualifier}ResetTermsAgreement]
+	@PortalId	INT
+AS
+	UPDATE {databaseOwner}{objectQualifier}UserPortals
```

Comment:
Isn't it irritation/misleading, if HasAgreedToTermsOn has a value but HasAgreedToTerms not?

Reply:
@sleupold I believe the reason to leave the value is that you can infer "last agreed" by this.

Reply:
IN this case, I would store "HasAgreedToTermsVersion" instead of "HasAgreedToTerms" and compare with PortalSetting "TermsVersion", which would be increased by the resetButton.